www/advanced-data-model.html

```diff
@@ -0,0 +1,621 @@
+    a default member method, we always say this with respect to 
+    <code>TileSource</code> class hierarchy.</b>
+    When a tile data is missing, an
+    <code>ImageLoader</code> instance gets new job assigned with
+    <code>ImageJob</code> instance. This instance receives all necessary
+    data extracted from methods of the <code>TileSource</code>, namely
+    <code>getTileUrl</code>, <code>getTilePostData</code>, <code>
+    getTileAjaxHeaders</code>. 
+</p>
+<p>        
+    The <code>ImageJob</code> instance then in return executes 
+    <code>downloadTileStart</code> method, also a member of 
+    <code>TileSource</code>. <code>ImageJob.prototype.finish()</code>
+    notifies the ImageLoader about the job completion and the data gets
+    pulled into the system.
+</p>
+<p>    
+    Last but not least, several other functions must be may be necessary
+    to override, especially if you change the default tile data type 
+    (<code>Image</code>).
+</p>
+<p>    
+    <code>getTileHashKey</code> should be overrided in cases your URL
+    does not uniquely distingish between different tile data. In other words,
+    if it is possible for two same tile URLs to contain different data, 
+    you should override this method. Usually this can happen when you
+    send the tile corrdinates in <b>POST</b> data. Usually, it is a good
+    idea to create the cache key based on tile coordinates, e.g. 
+    <code>level</code>, <code>x</code> and <code>y</code>. On the other hand,
+    if you know that some tiles <b>allways</b> have the same data although being 
+    on different positions, you might speed up the application by makeing the
+    key equal for such tiles.
+</p>
+<p>    
+    
+    Due to flexible ways of loading the data, <code>hasTransparency</code>
+    method should be also overrided to specify whether the certain tile source
+    contains trasparency. Usually the default implementation should work for you
+    (as all do), but with advanced features it might happen you experience 
+    artifacts during the viewing, like stacking of several tile images atop each
+    other and then vanishing when the viewer stops moving. In that case, use
+    this method to specify the trasparency flag correctly. 
+</p>
+
+<div class="demoarea">
+    <div class="demoheading">
+        Example: re-defining the tile downloading process
+    </div>
+    <div id="example-custom-tilesource-advanced" 
+         class="openseadragon" >
+    </div>
+    <p>
+        Below is an implementation of custom tile downloading.
+        A new <code>fetch</code> is used to distinguish between
+        the original implementation and the new approach.
+    </p>
+    <p>        
+        Furthermore, we use our freedom to handle failed requests
+        in different way. For documentation specifics, please, 
+        see JSDoc of corresponding functions.
+    </p>
+    <p>        
+        In this example, we override existing tile source implementation, 
```

Comment:
I don't agree. Inlining DZI means including, implementing unrelated/irrelevant stuff within the example code, e.g. handling DZI protocol. Overriding the functionality on the other hand emphasizes the independence of the approach: wanna use DZI and change how the data is fetched? Override these functions within the front-end implementation and it works! I doubt there are going to be any updates/changes to dzi in the future anyway since to my knowledge the protocol is no longer maintained.

How would you then prefer it to be done? Extend the DziTileSource class and replace it in the system? Still using the inline implementation?

Reply:
Hmm, that's a good point. We certainly want to model the smoothest approach, and doing it with in-line configuration requires a bunch of extra work.

I don't like directly overwriting DziTileSource, though, because it doesn't feel very clean. You shouldn't have to overwrite part of a library as part of using its API! I think it would be cleaner to subclass like so:


```
const MyDziTileSource = function(options) {
     OpenSeadragon.DziTileSource.apply( this, [ options ] );
};

OpenSeadragon.extend( MyDziTileSource.prototype, OpenSeadragon.DziTileSource.prototype, {
    ...
});
```

… but then you can't just give it a DZI URL and have it properly determine the type. What seems nice is if you could do something like:


```
viewer.open(new MyDziTileSource({ url: "/example-images/highsmith/highsmith.dzi" }));
```

… but that doesn't work with the current DziTileSource implementation.

So yeah, maybe what you've got is in fact the best way, given the current API.